### PR TITLE
Refactor Options to be immutable using a builder

### PR DESCRIPTION
**Pull request by varunu28**
**Why this change is required?**
Currently [Options.java](https://github.com/ebarlas/microhttp/blob/main/src/main/java/org/microhttp/Options.java) is effectively mutable and can be misused in the following way
The `EventLoop` [constructor](https://github.com/ebarlas/microhttp/blob/main/src/main/java/org/microhttp/EventLoop.java#L40) uses the options by reference. So while an options is passed to this constructor, it could also be updated later with any of the `with` methods such as [withHost](https://github.com/ebarlas/microhttp/blob/main/src/main/java/org/microhttp/Options.java#L58) resulting in unpredictable behavior.

**How does this change solves the above problem?**
This change aims to make the `Options` immutable using a builder `OptionsBuilder`. All the fields of `Options` are marked as `final` and an instance is only created after invoking the `build` method on the builder. Post this, none of the fields in `Options` are mutable. This prevents the code from ending up in any of the unpredictable scenarios where an option is updated post being passed to the event loop. 

A `getDefaultInstance` method is also added for use cases where we don't want to set any of the properties of `Options` and just use the default properties.

### CODE REVIEW

**Comment by varunu28** [comment]
@ebarlas This change is up for your review

**Comment by ebarlas** [comment]
Thanks for the contribution, @varunu28!

`Options` is very close to a record here. It might make sense to convert it to a record to avoid the boilerplate accessor methods.

Also, I wonder if pulling `OptionsBuilder` into its own source file would improve client code ever so slightly. It's a tad jarring to see `Options.OptionsBuilder.newBuilder()`.

**Comment by varunu28** [comment]
@ebarlas Thanks for the suggestion. Record seems like a good candidate for this case. The only concern I have is the long list of parameters which makes the code difficult to read.

To the point of separating `OptionsBuilder` into its own source file, with the current implementation the `build` method is able to access the private constructor of `Options` class. We can actually separate the class and have a `static` method under `Options` to build the instance using the builder eg `getInstanceUsingBuilder()`. This will ensure that `Options` is always created using the `Builder` object.

Let me know what you think.

**Comment by ebarlas** [comment]
In my rushed review I overlooked the tight relationship between `Options` and `OptionsBuilder`. The private constructor arrangement and the privileged access to the builder fields would both be lost with separate classes.

An `Options` record would relieve some of that tension, since the long parameter list is a reality and the builder can simply use that in the terminal `build`.

The long parameter list would effectively be out of sight for the programmer. And the record idiom would provide a nice form of self documentation in terms of immutability, data access, etc.

I think it's worth exploring.

**Comment by varunu28** [comment]
I looked into records(No pun intended). Records do pose an issue in this use case as not all the fields are required for `Options` and Record by default provide an all-args constructor. So we will need to have custom constructors for all possible combinations of input parameters.

More details: https://www.vojtechruzicka.com/java-records/#custom-constructors

Unless we force the client to provide all parameters, we will have to add a lot of boilerplate code for custom constructors.

Though we can separate the builder class to a separate source file, making `Options` cleaner. This will require an additional method to build the class instance as part of `Options` class

**Comment by ebarlas** [comment]
The builder _already_ has defaults for all fields. So when `build` is called, it will alway pass all of them to the `Options` constructor.

Clients would simply use the builder.

**Comment by varunu28** [comment]
Oh I see. So you are thinking about a combination of both builder and the record. Neat idea. 
I will go ahead and refactor the code to do accomodate below changes:

- Make the `Options` class as a `Record`
- Move the builder class to a separate source file.

**Comment by varunu28** [comment]
> Oh I see. So you are thinking about a combination of both builder and the record. Neat idea. I will go ahead and refactor the code to do accomodate below changes:
> 
> * Make the `Options` class as a `Record`
> * Move the builder class to a separate source file.

Refactored the change to fulfill above requirements.

**Comment by ebarlas** [comment]
Looks great!

Last question: why do we need `OptionsBuilder.newBuilder()`? `OptionsBuilder` can simply have a public constructor.

**Comment by varunu28** [comment]
TL;DR Both the method and the public constructor solve the purpose. Just a matter of preference

Both achieve the same purpose. It is just I have a preference of having `static` methods over public constructors for class instantiation. It ends up being more readable and extendible if we plan on having multiple constructors in future for the builder. It is also in line with semantics of `builder` used in protocol buffers.

**Comment by ebarlas** [comment]
Agreed. I know static methods are a common builder idiom. Merging. Thanks again!

**Comment by varunu28** [comment]
Thanks @ebarlas . I really like the project and discovered it through a reddit thread. I have started playing with it and will love to contribute more in the coming days.
